components/merkledb/src/proto/mod.rs

```diff
@@ -17,44 +17,143 @@
 // For protobuf generated files.
 #![allow(bare_trait_objects)]
 
-pub use self::list_proof::*;
+use failure::Error;
+use protobuf::{well_known_types::Empty, RepeatedField};
+use std::borrow::Cow;
+
+use crate::{proof_map_index::ProofPath, BinaryKey, BinaryValue};
 use exonum_crypto::proto::*;
+use exonum_proto::ProtobufConvert;
+
+pub use self::{list_proof::*, map_proof::*};
 
 include!(concat!(env!("OUT_DIR"), "/protobuf_mod.rs"));
 
+impl<K, V> ProtobufConvert for crate::MapProof<K, V>
+where
+    K: BinaryKey + ToOwned<Owned = K>,
```

Comment:
Could it make sense to have `K: BinaryKey + BinaryValue + ...` here? That is, the key type has both key and value implementations and these implementations agree. To expand on this idea, there may be a marker trait to express `BinaryKey` and `BinaryValue` agreement (say, `BinaryKeyFromValue`); `impl_binary_key_for_binary_value` macro can implement this trait. Further, it could be worth investigating whether `impl<T: BinaryValue + BinaryKeyFromValue> BinaryKey for T` could work. 

This could be beneficial if we think that client code can "monomorphize" `MapProof` definition by replacing `bytes` with the target message type. If we don't think that this is a viable use case, `K` constraints could be left as is (but then a question about `BinaryKey` interface panicking instead of emitting errors remains.)

Reply:
I prefer current variant. I think it's better to rewrite `BinaryKey` definition.

Reply:
@dmitry-timofeev To hear feedback from our Java overlords:

- Does EJB have any plans to "monomorphize" map proofs (e.g., for the state hash aggregator)?
- Would you be OK with the serialization of keys in `MapProof` being potentially not correct Protobuf? (Mind you, this may happen for a user-defined type, not only for integers, if the user is interested in keys being ordered in a particular way.)

Reply:
I am not sure I understand what you mean by 'monomorphizing' the MapProof. As MapProof message is currently a message defined by Exonum and the protobuf does not support generics (type parameters), than this process requires the user creating a separate message to represent a proof (with no support from the framework with verifying that)? Or are you considering some internal representation of a proof decoding the keys/values given the user provides the expected types?

Regarding the 2nd question, if they have appropriate type (just bytes?), then I don't think it will be a problem. It is also currently possible to supply non-protobuf-serialized keys/values in Java.